### PR TITLE
Add RunnerPipelineoptionsFactory and Registrar to support loading PipelineOptionsFactory in runtime

### PR DESCRIPTION
**Pull request by xinyuiscool**
As part of our clean up, we want to make the API consistent with open source and agnostic of runner-specific PipelineOptions. Adding this interface and Registrar to load runner-specific PipelineOptionsFactory during runtime, e.g. SamzaPipelineOptionsFactory. 

Tested by beam-runner MP and local deployed jobs. Also fixed the problem in local snapshot publishing of beam due to bom. I changed the BeamModulePlugin to generate pom.xml without bom dependencies.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -1827,7 +1827,8 @@ class BeamModulePlugin implements Plugin<Project> {
                     configuration: (configuration.shadowClosure ? 'shadow' : 'implementation'), scope: 'compile')
                 generateDependenciesFromConfiguration(configuration: 'provided', scope: 'provided')
 
-                if (!boms.isEmpty()) {
+                // LinkedIn: no boms in pom.xml so it can be used in MPs
```
Comment (yananhao12) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Maybe also mention in the comment that we only need to skip boms for LinkedIn Snapshot build?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/RunnerPipelineOptionsFactory.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  /**
+   * Registrar of the {@link RunnerPipelineOptionsFactory}. Must be a single class in a runner's
+   * runtime.
+   */
+  interface Registrar {
+    RunnerPipelineOptionsFactory create();
+  }
+
+  /**
+   * Fing the {@link RunnerPipelineOptionsFactory} to load runner-specific {@link PipelineOptions}.
```
Comment (yananhao12) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/RunnerPipelineOptionsFactory.java]: typo "Fing"?

**Review by yananhao12 — COMMENTED** [review]
LGTM, just minor comment

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 1 above)
Comment (xinyuiscool) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Makes sense. Added more comments.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/RunnerPipelineOptionsFactory.java (hunk 2 above)
Comment (xinyuiscool) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/RunnerPipelineOptionsFactory.java]: Good catch. Fixed.

**Review by yananhao12 — APPROVED** [review]

**Pull request merged**